applications/callflow/src/module/cf_branch_property.erl

```diff
@@ -0,0 +1,92 @@
+%%%-------------------------------------------------------------------
+%%% @copyright (C) 2013-2016, 2600Hz INC
+%%% @doc
+%%%
+%%% "data":{
+%%%   "action":"branch", % the only one action supported currently
+%%%   "scope": "[device|user|account]",
+%%%   "property": "whatever_field_you_want"
+%%% }
+%%%
+%%% Branch to child whose name pointed in callflow's data section
+%%%
+%%% @end
+%%% @contributors
+%%%   SIPLABS LLC (Maksim Krzhemenevskiy)
+%%%-------------------------------------------------------------------
+-module(cf_branch_property).
+
+-include("../callflow.hrl").
+
+%% API
+-export([handle/2]).
+
+-spec handle(wh_json:object(), whapps_call:call()) -> 'ok'.
+handle(Data, Call) ->
+    Scope = wh_json:get_binary_value(<<"scope">>, Data),
+    PointedField = wh_json:get_binary_value(<<"property">>, Data),
+    ChildName = case is_binary(PointedField) of
+                    'true' -> lookup_child_name(Scope, Call, PointedField);
+                    'false' ->
+                        lager:debug("value for \"property\" is unset"),
+                        'undefined'
+                end,
+    case is_binary(ChildName) of
+        'true' -> cf_exe:continue(ChildName, Call);
+        'false' -> cf_exe:continue(Call)
+    end.
+
+-spec lookup_child_name(api_binary(), whapps_call:call(), ne_binary()) -> api_binary().
+lookup_child_name(Scope, Call, Field) ->
+    case Scope of
+        <<"device">> -> targeted_lookup(Scope, Call, Field);
+        <<"user">> -> targeted_lookup(Scope, Call, Field);
+        <<"account">> -> targeted_lookup(Scope, Call, Field);
+        _ -> automatic_lookup(Call, Field)
+    end.
+
+-spec automatic_lookup(whapps_call:call(), ne_binary()) -> api_binary().
+automatic_lookup(Call, Field) ->
+    case cf_endpoint:get(Call) of
+        {'ok', JObj} -> wh_json:get_value(Field, JObj);
+        _Else ->
+            lager:debug("can not lookup endpoint for caller"),
+            'undefined'
+    end.
+
+-spec targeted_lookup(ne_binary(), whapps_call:call(), ne_binary()) -> api_binary().
+targeted_lookup(Scope, Call, Field) ->
+    AuthId = whapps_call:authorizing_id(Call),
+    DocId = case {Scope, whapps_call:authorizing_type(Call)} of
+                      {<<"account">>, _} -> whapps_call:account_id(Call);
+                      {Same, Same} -> AuthId;
```

Comment:
If you pattern match as above, this could be a second function pattern match:

```erlang
targeted_lookup(Scope, Call, Field) ->
    ChildDocId = targeted_lookup(Scope, Call, Field, whapps_call:authorizing_id(Call)),
    lookup_doc(Scope, Call, Field, ChildDocId).

targeted_lookup(Scope, Call, Field, Scope) -> whapps_call:authorizing_id(Call);
targeted_lookup(_Scope, Call, _Field, <<"device">>) -> lookup_device_owner(Call, whapps_call:authorizing_id(Call));
targeted_lookup(_Scope, _Call, _Field, <<"user">>) -> 'undefined';
targeted_lookup(_Scope, _Call, _Field, _AuthType) ->
    lager:debug("unsupported authorizing type: ~s", [_AuthType]),
    'undefined'.

lookup_doc(_Scope, _Call, _Field, 'undefined') ->
    lager:debug("cannot find document for current scope: ~s", [_Scope]),
    'undefined';
lookup_doc(_Scope, Call, Field, DocId) ->
    lookup_childname(Call, Field, DocId).
```

Also, I prefer it when function arguments aren't re-arranged for no reason. `lookup_childname/3` in your code should follow the order of arguments from the calling function; I think it reads better and the compiler may be able to optimize the callstack a little(?) but mostly I find it more consistent.